File: 프로그래머스/lv3/여행경로-dfs.py
# DFS 재귀 방식
# 처음에 출발하는 항공, 거쳐가는 경로를 dfs에 넣어
# 출발하는 항공이 같고 아직 방문하지 않은 항공권이면
# 해당 항공권의 도착지, 거쳐가는 경로를 넣어 깊이 탐색을 진행한다.
# 거쳐가는 경로의 길이가 티켓 길이 + 1이면 모든 티켓을 사용하게 된 것이므로 정답 리스트에 해당 경로를 넣어준다.
# 이렇게 모든 경로들이 저장된 정답 리스트에서 알파벳이 작은 기준으로 정렬하게하여 
# 제일 첫 원소의 경로가 알파벳이 작은 순서의 경로이다. 

def solution(tickets):
    answer = []
    
    m = len(tickets)
    visited = [False] * m
    
    def dfs(start, path):
        if len(path) == m+1:                      # 거쳐가는 경로의 길이가 티켓 길이 + 1이면
            answer.append(path)                   # 모든 티켓을 사용하게 된 것이므로 정답 리스트에 해당 경로를 넣어준다.
            return
        
        
        for idx, ticket in enumerate(tickets):
            if start == ticket[0] and visited[idx] == False:    # 출발하는 항공이 같고 아직 방문하지 않은 항공권이면
                visited[idx] = True                             # 방문 처리를 해주고
                dfs(ticket[1], path + [ticket[1]])              # 해당 항공권의 도착지, 거쳐가는 경로를 넣어 깊이 탐색을 진행한다.
                visited[idx] = False                            # 끝까지 깊이 탐색이 완료되었으면 방문 처리 했던 것을 백트랙킹 해준다.
        
        
    dfs("ICN", ["ICN"])                           # 처음에 출발하는 항공, 거쳐가는 경로를 dfs에 넣고 진행
    
    
    answer.sort()                                 # 이렇게 모든 경로들이 저장된 정답 리스트에서 알파벳이 작은 기준으로 정렬하게하여 
   
    return answer[0]                              # 제일 첫 원소의 경로가 알파벳이 작은 순서의 경로이다. 

File: 자바/dfs,bfs/여행경로-dfs.java
import java.util.*;
class Solution {

    public String[] solution(String[][] tickets) {
        List<String> answer = new ArrayList<>();
        boolean[] visited = new boolean[tickets.length];

        dfs("ICN", "ICN", tickets, answer, visited, 0);

        Collections.sort(answer);           // 여행 경로가 다양하게 나오는데 이중 알파벳순이라고 했으므로 정렬

        return answer.get(0).split(" ");    // "ICN SFO ATL" 같이 공백으로 구분했고 알바벳순으로 정렬한 제일 앞의 여행경로가 정답이므로
                                            //  첫번째항의 각 공백을 기준으로 String 배열로 변환
    }


    private void dfs(String start, String path, String[][] tickets, List<String> answer, boolean[] visited, int depth) {
        if (depth == tickets.length) {      // 주어진 티켓을 모두 조회했다면
            answer.add(path);               // 현재까지의 여행 경로를 정답 리스트에 추가하고 반환
            return;
        }

        for (int i = 0; i < tickets.length; i++) {
            if (start.equals(tickets[i][0]) && !visited[i]) {
                visited[i] = true;  // 방문했음으로 설정하고
                dfs(tickets[i][1], path + " " + tickets[i][1], tickets, answer, visited, depth + 1); // 재귀를 하고
                visited[i] = false; // for문 다음 조회에서 방문한 티켓을 다시 방문할 수 있어야하므로 방문안함으로 다시 설정
            }
        }
    }
}